File: OTCDerivativesCalculatorModule/Project_CSharp/XmlBuilder/XmlSerializedClass/fpml-business-events-5-4/TradeReferenceInformation.cs
using System;
using System.Text;
using System.Collections.Generic;
using System.Xml;

namespace FpmlSerializedCSharp
{
    public class TradeReferenceInformation : ISerialized
    {
        public TradeReferenceInformation(XmlNode xmlNode)
        {
            XmlNode originatingEventNode = xmlNode.SelectSingleNode("originatingEvent");
            
            if (originatingEventNode != null)
            {
                if (originatingEventNode.Attributes["href"] != null || originatingEventNode.Attributes["id"] != null) 
                {
                    if (originatingEventNode.Attributes["id"] != null) 
                    {
                        originatingEventIDRef_ = originatingEventNode.Attributes["id"].Value;
                        OriginatingEvent ob = new OriginatingEvent(originatingEventNode);
                        IDManager.SetID(originatingEventIDRef_, ob);
                    }
                    else if (originatingEventNode.Attributes["href"] != null)
                    {
                        originatingEventIDRef_ = originatingEventNode.Attributes["href"].Value;
                    }
                    else
                    {
                        originatingEvent_ = new OriginatingEvent(originatingEventNode);
                    }
                }
                else
                {
                    originatingEvent_ = new OriginatingEvent(originatingEventNode);
                }
            }
            
        
            XmlNode terminatingEventNode = xmlNode.SelectSingleNode("terminatingEvent");
            
            if (terminatingEventNode != null)
            {
                if (terminatingEventNode.Attributes["href"] != null || terminatingEventNode.Attributes["id"] != null) 
                {
                    if (terminatingEventNode.Attributes["id"] != null) 
                    {
                        terminatingEventIDRef_ = terminatingEventNode.Attributes["id"].Value;
                        TerminatingEvent ob = new TerminatingEvent(terminatingEventNode);
                        IDManager.SetID(terminatingEventIDRef_, ob);
                    }
                    else if (terminatingEventNode.Attributes["href"] != null)
                    {
                        terminatingEventIDRef_ = terminatingEventNode.Attributes["href"].Value;
                    }
                    else
                    {
                        terminatingEvent_ = new TerminatingEvent(terminatingEventNode);
                    }
                }
                else
                {
                    terminatingEvent_ = new TerminatingEvent(terminatingEventNode);
                }
            }
            
        
            XmlNodeList partyTradeIdentifierNodeList = xmlNode.SelectNodes("partyTradeIdentifier");
            
            if (partyTradeIdentifierNodeList != null)
            {
                this.partyTradeIdentifier_ = new List<PartyTradeIdentifier>();
                foreach (XmlNode item in partyTradeIdentifierNodeList)
                {
                    if (item.Attributes["href"] != null || item.Attributes["id"] != null) 
                    {
                        if (item.Attributes["id"] != null) 
                        {
                            partyTradeIdentifierIDRef_ = item.Attributes["id"].Value;
                            partyTradeIdentifier_.Add(new PartyTradeIdentifier(item));
                            IDManager.SetID(partyTradeIdentifierIDRef_, partyTradeIdentifier_[partyTradeIdentifier_.Count - 1 ]);
                        }
                        else if (item.Attributes["href"] != null)
                        {
                            partyTradeIdentifierIDRef_ = item.Attributes["href"].Value;
                        }
                        else
                        {
                        partyTradeIdentifier_.Add(new PartyTradeIdentifier(item));
                        }
                    }
                    else
                    {
                        partyTradeIdentifier_.Add(new PartyTradeIdentifier(item));
                    }
                }
            }
            
        
            XmlNodeList partyTradeInformationNodeList = xmlNode.SelectNodes("partyTradeInformation");
            
            if (partyTradeInformationNodeList != null)
            {
                this.partyTradeInformation_ = new List<PartyTradeInformation>();
                foreach (XmlNode item in partyTradeInformationNodeList)
                {
                    if (item.Attributes["href"] != null || item.Attributes["id"] != null) 
                    {
                        if (item.Attributes["id"] != null) 
                        {
                            partyTradeInformationIDRef_ = item.Attributes["id"].Value;
                            partyTradeInformation_.Add(new PartyTradeInformation(item));
                            IDManager.SetID(partyTradeInformationIDRef_, partyTradeInformation_[partyTradeInformation_.Count - 1 ]);
                        }
                        else if (item.Attributes["href"] != null)
                        {
                            partyTradeInformationIDRef_ = item.Attributes["href"].Value;
                        }
                        else
                        {
                        partyTradeInformation_.Add(new PartyTradeInformation(item));
                        }
                    }
                    else
                    {
                        partyTradeInformation_.Add(new PartyTradeInformation(item));
                    }
                }
            }
            
        
            XmlNode productTypeNode = xmlNode.SelectSingleNode("productType");
            
            if (productTypeNode != null)
            {
                if (productTypeNode.Attributes["href"] != null || productTypeNode.Attributes["id"] != null) 
                {
                    if (productTypeNode.Attributes["id"] != null) 
                    {
                        productTypeIDRef_ = productTypeNode.Attributes["id"].Value;
                        ProductType ob = new ProductType(productTypeNode);
                        IDManager.SetID(productTypeIDRef_, ob);
                    }
                    else if (productTypeNode.Attributes["href"] != null)
                    {
                        productTypeIDRef_ = productTypeNode.Attributes["href"].Value;
                    }
                    else
                    {
                        productType_ = new ProductType(productTypeNode);
                    }
                }
                else
                {
                    productType_ = new ProductType(productTypeNode);
                }
            }
            
        
            XmlNode productIdNode = xmlNode.SelectSingleNode("productId");
            
            if (productIdNode != null)
            {
                if (productIdNode.Attributes["href"] != null || productIdNode.Attributes["id"] != null) 
                {
                    if (productIdNode.Attributes["id"] != null) 
                    {
                        productIdIDRef_ = productIdNode.Attributes["id"].Value;
                        ProductId ob = new ProductId(productIdNode);
                        IDManager.SetID(productIdIDRef_, ob);
                    }
                    else if (productIdNode.Attributes["href"] != null)
                    {
                        productIdIDRef_ = productIdNode.Attributes["href"].Value;
                    }
                    else
                    {
                        productId_ = new ProductId(productIdNode);
                    }
                }
                else
                {
                    productId_ = new ProductId(productIdNode);
                }
            }
            
        
        }
        
    
        #region OriginatingEvent_
        private OriginatingEvent originatingEvent_;
        public OriginatingEvent OriginatingEvent_
        {
            get
            {
                if (this.originatingEvent_ != null)
                {
                    return this.originatingEvent_; 
                }
                else if (this.originatingEventIDRef_ != null)
                {
                    originatingEvent_ = IDManager.getID(originatingEventIDRef_) as OriginatingEvent;
                    return this.originatingEvent_; 
                }
                else
                {
                      return this.originatingEvent_; 
                }
            }
            set
            {
                if (this.originatingEvent_ != value)
                {
                    this.originatingEvent_ = value;
                }
            }
        }
        #endregion
        
        public string originatingEventIDRef_ { get; set; }
        #region TerminatingEvent_
        private TerminatingEvent terminatingEvent_;
        public TerminatingEvent TerminatingEvent_
        {
            get
            {
                if (this.terminatingEvent_ != null)
                {
                    return this.terminatingEvent_; 
                }
                else if (this.terminatingEventIDRef_ != null)
                {
                    terminatingEvent_ = IDManager.getID(terminatingEventIDRef_) as TerminatingEvent;
                    return this.terminatingEvent_; 
                }
                else
                {
                      return this.terminatingEvent_; 
                }
            }
            set
            {
                if (this.terminatingEvent_ != value)
                {
                    this.terminatingEvent_ = value;
                }
            }
        }
        #endregion
        
        public string terminatingEventIDRef_ { get; set; }
        #region PartyTradeIdentifier_
        private List<PartyTradeIdentifier> partyTradeIdentifier_;
        public List<PartyTradeIdentifier> PartyTradeIdentifier_
        {
            get
            {
                if (this.partyTradeIdentifier_ != null)
                {
                    return this.partyTradeIdentifier_; 
                }
                else if (this.partyTradeIdentifierIDRef_ != null)
                {
                    return this.partyTradeIdentifier_; 
                }
                else
                {
                      return this.partyTradeIdentifier_; 
                }
            }
            set
            {
                if (this.partyTradeIdentifier_ != value)
                {
                    this.partyTradeIdentifier_ = value;
                }
            }
        }
        #endregion
        
        public string partyTradeIdentifierIDRef_ { get; set; }
        #region PartyTradeInformation_
        private List<PartyTradeInformation> partyTradeInformation_;
        public List<PartyTradeInformation> PartyTradeInformation_
        {
            get
            {
                if (this.partyTradeInformation_ != null)
                {
                    return this.partyTradeInformation_; 
                }
                else if (this.partyTradeInformationIDRef_ != null)
                {
                    return this.partyTradeInformation_; 
                }
                else
                {
                      return this.partyTradeInformation_; 
                }
            }
            set
            {
                if (this.partyTradeInformation_ != value)
                {
                    this.partyTradeInformation_ = value;
                }
            }
        }
        #endregion
        
        public string partyTradeInformationIDRef_ { get; set; }
        #region ProductType_
        private ProductType productType_;
        public ProductType ProductType_
        {
            get
            {
                if (this.productType_ != null)
                {
                    return this.productType_; 
                }
                else if (this.productTypeIDRef_ != null)
                {
                    productType_ = IDManager.getID(productTypeIDRef_) as ProductType;
                    return this.productType_; 
                }
                else
                {
                      return this.productType_; 
                }
            }
            set
            {
                if (this.productType_ != value)
                {
                    this.productType_ = value;
                }
            }
        }
        #endregion
        
        public string productTypeIDRef_ { get; set; }
        #region ProductId_
        private ProductId productId_;
        public ProductId ProductId_
        {
            get
            {
                if (this.productId_ != null)
                {
                    return this.productId_; 
                }
                else if (this.productIdIDRef_ != null)
                {
                    productId_ = IDManager.getID(productIdIDRef_) as ProductId;
                    return this.productId_; 
                }
                else
                {
                      return this.productId_; 
                }
            }
            set
            {
                if (this.productId_ != value)
                {
                    this.productId_ = value;
                }
            }
        }
        #endregion
        
        public string productIdIDRef_ { get; set; }
        public string choiceStr_0;
        
    
        
    
    }
    
}

File: OTCDerivativesCalculatorModule/Project_Cpp/lib_static/FpmlSerialized/GenClass/fpml-business-events-5-4/TradeReferenceInformation.cpp
// TradeReferenceInformation.cpp 
#include "TradeReferenceInformation.hpp"
#ifdef ConsolePrint
    #include <iostream>
#endif 

namespace FpmlSerialized {

TradeReferenceInformation::TradeReferenceInformation(TiXmlNode* xmlNode)
: ISerialized(xmlNode)
{
    #ifdef ConsolePrint
        std::string initialtap_ = FileManager::instance().tap_;
        FileManager::instance().tap_.append("   ");
    #endif 
   //originatingEventNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* originatingEventNode = xmlNode->FirstChildElement("originatingEvent");

   if(originatingEventNode){originatingEventIsNull_ = false;}
   else{originatingEventIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- originatingEventNode , address : " << originatingEventNode << std::endl;
   #endif
   if(originatingEventNode)
   {
      if(originatingEventNode->Attribute("href") || originatingEventNode->Attribute("id"))
      {
          if(originatingEventNode->Attribute("id"))
          {
             originatingEventIDRef_ = originatingEventNode->Attribute("id");
             originatingEvent_ = boost::shared_ptr<OriginatingEvent>(new OriginatingEvent(originatingEventNode));
             originatingEvent_->setID(originatingEventIDRef_);
             IDManager::instance().setID(originatingEventIDRef_,originatingEvent_);
          }
          else if(originatingEventNode->Attribute("href")) { originatingEventIDRef_ = originatingEventNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { originatingEvent_ = boost::shared_ptr<OriginatingEvent>(new OriginatingEvent(originatingEventNode));}
   }

   //terminatingEventNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* terminatingEventNode = xmlNode->FirstChildElement("terminatingEvent");

   if(terminatingEventNode){terminatingEventIsNull_ = false;}
   else{terminatingEventIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- terminatingEventNode , address : " << terminatingEventNode << std::endl;
   #endif
   if(terminatingEventNode)
   {
      if(terminatingEventNode->Attribute("href") || terminatingEventNode->Attribute("id"))
      {
          if(terminatingEventNode->Attribute("id"))
          {
             terminatingEventIDRef_ = terminatingEventNode->Attribute("id");
             terminatingEvent_ = boost::shared_ptr<TerminatingEvent>(new TerminatingEvent(terminatingEventNode));
             terminatingEvent_->setID(terminatingEventIDRef_);
             IDManager::instance().setID(terminatingEventIDRef_,terminatingEvent_);
          }
          else if(terminatingEventNode->Attribute("href")) { terminatingEventIDRef_ = terminatingEventNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { terminatingEvent_ = boost::shared_ptr<TerminatingEvent>(new TerminatingEvent(terminatingEventNode));}
   }

   //partyTradeIdentifierNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* partyTradeIdentifierNode = xmlNode->FirstChildElement("partyTradeIdentifier");

   if(partyTradeIdentifierNode){partyTradeIdentifierIsNull_ = false;}
   else{partyTradeIdentifierIsNull_ = true;}

   if(partyTradeIdentifierNode)
   {
      for(partyTradeIdentifierNode; partyTradeIdentifierNode; partyTradeIdentifierNode = partyTradeIdentifierNode->NextSiblingElement("partyTradeIdentifier")){
          #ifdef ConsolePrint
              FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- partyTradeIdentifierNode , address : " << partyTradeIdentifierNode << std::endl;
          #endif
          if(partyTradeIdentifierNode->Attribute("href") || partyTradeIdentifierNode->Attribute("id"))
          {
              if(partyTradeIdentifierNode->Attribute("id"))
              {
                  partyTradeIdentifierIDRef_ = partyTradeIdentifierNode->Attribute("id");
                  partyTradeIdentifier_.push_back(boost::shared_ptr<PartyTradeIdentifier>(new PartyTradeIdentifier(partyTradeIdentifierNode)));
                  partyTradeIdentifier_.back()->setID(partyTradeIdentifierIDRef_);
                  IDManager::instance().setID(partyTradeIdentifierIDRef_, partyTradeIdentifier_.back());
              }
              else if(partyTradeIdentifierNode->Attribute("href")) { partyTradeIdentifierIDRef_ = partyTradeIdentifierNode->Attribute("href");}
              else { QL_FAIL("id or href error"); }
          }
          else { partyTradeIdentifier_.push_back(boost::shared_ptr<PartyTradeIdentifier>(new PartyTradeIdentifier(partyTradeIdentifierNode)));}
      }
   }
   else {
       #ifdef ConsolePrint
           FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- partyTradeIdentifierNode , address : " << partyTradeIdentifierNode << std::endl;
       #endif
   }

   //partyTradeInformationNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* partyTradeInformationNode = xmlNode->FirstChildElement("partyTradeInformation");

   if(partyTradeInformationNode){partyTradeInformationIsNull_ = false;}
   else{partyTradeInformationIsNull_ = true;}

   if(partyTradeInformationNode)
   {
      for(partyTradeInformationNode; partyTradeInformationNode; partyTradeInformationNode = partyTradeInformationNode->NextSiblingElement("partyTradeInformation")){
          #ifdef ConsolePrint
              FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- partyTradeInformationNode , address : " << partyTradeInformationNode << std::endl;
          #endif
          if(partyTradeInformationNode->Attribute("href") || partyTradeInformationNode->Attribute("id"))
          {
              if(partyTradeInformationNode->Attribute("id"))
              {
                  partyTradeInformationIDRef_ = partyTradeInformationNode->Attribute("id");
                  partyTradeInformation_.push_back(boost::shared_ptr<PartyTradeInformation>(new PartyTradeInformation(partyTradeInformationNode)));
                  partyTradeInformation_.back()->setID(partyTradeInformationIDRef_);
                  IDManager::instance().setID(partyTradeInformationIDRef_, partyTradeInformation_.back());
              }
              else if(partyTradeInformationNode->Attribute("href")) { partyTradeInformationIDRef_ = partyTradeInformationNode->Attribute("href");}
              else { QL_FAIL("id or href error"); }
          }
          else { partyTradeInformation_.push_back(boost::shared_ptr<PartyTradeInformation>(new PartyTradeInformation(partyTradeInformationNode)));}
      }
   }
   else {
       #ifdef ConsolePrint
           FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- partyTradeInformationNode , address : " << partyTradeInformationNode << std::endl;
       #endif
   }

   //productTypeNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* productTypeNode = xmlNode->FirstChildElement("productType");

   if(productTypeNode){productTypeIsNull_ = false;}
   else{productTypeIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- productTypeNode , address : " << productTypeNode << std::endl;
   #endif
   if(productTypeNode)
   {
      if(productTypeNode->Attribute("href") || productTypeNode->Attribute("id"))
      {
          if(productTypeNode->Attribute("id"))
          {
             productTypeIDRef_ = productTypeNode->Attribute("id");
             productType_ = boost::shared_ptr<ProductType>(new ProductType(productTypeNode));
             productType_->setID(productTypeIDRef_);
             IDManager::instance().setID(productTypeIDRef_,productType_);
          }
          else if(productTypeNode->Attribute("href")) { productTypeIDRef_ = productTypeNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { productType_ = boost::shared_ptr<ProductType>(new ProductType(productTypeNode));}
   }

   //productIdNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* productIdNode = xmlNode->FirstChildElement("productId");

   if(productIdNode){productIdIsNull_ = false;}
   else{productIdIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- productIdNode , address : " << productIdNode << std::endl;
   #endif
   if(productIdNode)
   {
      if(productIdNode->Attribute("href") || productIdNode->Attribute("id"))
      {
          if(productIdNode->Attribute("id"))
          {
             productIdIDRef_ = productIdNode->Attribute("id");
             productId_ = boost::shared_ptr<ProductId>(new ProductId(productIdNode));
             productId_->setID(productIdIDRef_);
             IDManager::instance().setID(productIdIDRef_,productId_);
          }
          else if(productIdNode->Attribute("href")) { productIdIDRef_ = productIdNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { productId_ = boost::shared_ptr<ProductId>(new ProductId(productIdNode));}
   }

    #ifdef ConsolePrint
        FileManager::instance().tap_ = initialtap_;
    #endif 
}
boost::shared_ptr<OriginatingEvent> TradeReferenceInformation::getOriginatingEvent()
{
   if(!this->originatingEventIsNull_){
        if(originatingEventIDRef_ != ""){
             return boost::shared_static_cast<OriginatingEvent>(IDManager::instance().getID(originatingEventIDRef_));
        }else{
             return this->originatingEvent_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<OriginatingEvent>();
   }
}
boost::shared_ptr<TerminatingEvent> TradeReferenceInformation::getTerminatingEvent()
{
   if(!this->terminatingEventIsNull_){
        if(terminatingEventIDRef_ != ""){
             return boost::shared_static_cast<TerminatingEvent>(IDManager::instance().getID(terminatingEventIDRef_));
        }else{
             return this->terminatingEvent_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<TerminatingEvent>();
   }
}
std::vector<boost::shared_ptr<PartyTradeIdentifier>> TradeReferenceInformation::getPartyTradeIdentifier()
{
   if(!this->partyTradeIdentifierIsNull_){
        if(partyTradeIdentifierIDRef_ != ""){
             return this->partyTradeIdentifier_;
        }else{
             return this->partyTradeIdentifier_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return std::vector<boost::shared_ptr<PartyTradeIdentifier>>();
   }
}
std::vector<boost::shared_ptr<PartyTradeInformation>> TradeReferenceInformation::getPartyTradeInformation()
{
   if(!this->partyTradeInformationIsNull_){
        if(partyTradeInformationIDRef_ != ""){
             return this->partyTradeInformation_;
        }else{
             return this->partyTradeInformation_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return std::vector<boost::shared_ptr<PartyTradeInformation>>();
   }
}
boost::shared_ptr<ProductType> TradeReferenceInformation::getProductType()
{
   if(!this->productTypeIsNull_){
        if(productTypeIDRef_ != ""){
             return boost::shared_static_cast<ProductType>(IDManager::instance().getID(productTypeIDRef_));
        }else{
             return this->productType_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<ProductType>();
   }
}
boost::shared_ptr<ProductId> TradeReferenceInformation::getProductId()
{
   if(!this->productIdIsNull_){
        if(productIdIDRef_ != ""){
             return boost::shared_static_cast<ProductId>(IDManager::instance().getID(productIdIDRef_));
        }else{
             return this->productId_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<ProductId>();
   }
}
}

